[16.0][IMP] account_reconcile_oca: Improve multicurrency management

etobella opened the pull request:
Tries to fix #692 #693 
@florian-dacosta Can you check it please?

etobella committed:
[IMP] account_reconcile_oca: Improve multicurrency management.

We will use currency of the line in order to get the suspense and max line value

florian-dacosta commented:
@etobella Thanks for the PR!
Well, it fix the example I gave in the issue yes, thanks, but it does not fix all cases, like the one in the second issue.
Let me show you what I have with the same setup from https://github.com/OCA/account-reconcile/issues/693

Default screen when going to reconcile UI, we have a rounding issue
![1-bfore-reconcile](https://github.com/user-attachments/assets/93bd9fac-a950-45df-aac6-e9e1149dfaee)

After entering the 121000 account in manual tab : 
![2-accounting-entry](https://github.com/user-attachments/assets/a42511c3-e96e-4331-9ce1-13e321dafc44)

etobella commented:
@florian-dacosta Round 2
![image](https://github.com/user-attachments/assets/78286dc8-a754-478b-8008-8cac2955afcf)
I added a rounding factor in order to remove this possible error (the rounding should only be used on comparison)

florian-dacosta commented:
@etobella I still don't get why there is a rounding to find an amount that is written in the `amount` field of the `account.bank.statement.line`. Why the 259 200 needs to be computed  and not read for the line ?

Anyway, I've tested the later version
I still get this rounding issue before chosing the invoice.
And once validated, the state /payment on the invoice is not ok.

Go to reconcile UI - Not OK
![1-rec-ui](https://github.com/user-attachments/assets/001f7dda-6f7f-482b-b5e6-f82fb9290f25)

Choose the invoice - OK
![2-after-choosing-invoice](https://github.com/user-attachments/assets/cd13c5c8-8846-4e4f-9040-617705e4b3d8)

Reconciled invoice after validation - Not OK
![3-invoice-partial](https://github.com/user-attachments/assets/f1ec6359-c1f5-40eb-a4fe-161fa31c3b40)

The bank accounting entry is missing the currency on the 121000 line
![4-accounting-entry](https://github.com/user-attachments/assets/2d398d60-8d21-4bce-948a-56a4689541bc)

etobella commented:
We have a rounding issue because the reconcile is done on the currency of the company. Also we cannot use that amount, because we might have previous lines.

The only option I can see is to do the reconcile in the amount of the statement line.

This is more work to do and I just wanted to add a patch, but seems we need to do a real refactoring (it is good to have so many tests)

PaulGoubert commented:
Thank you for your help with these issues, @etobella.
As previously mentioned by @florian-dacosta, we're in a rush on this topic due to our customer's situation—they've been expecting to use it properly for 6 months now. I hope you can tackle these topics and help us resolve the issues.

etobella commented:
@PaulGoubert @florian-dacosta Round 3

I have changed the flow in order to compare using the currency of the line. It has no sense to move from EUR to USD to EUR, so it should remove some of this errors.

etobella commented:
The only issue I found is that it is not showing properly when reconciling with the Foraign Exchange Loss

Before Reconciling
![image](https://github.com/user-attachments/assets/516ba435-557b-4bc6-85d6-e0d80ecb830a)
 
After reconciling
![image](https://github.com/user-attachments/assets/bfce31f3-f323-42fd-8a7c-f23fdfb3bf1e)

The foreign loss has been lost, but it is correctly reconciled 
![image](https://github.com/user-attachments/assets/f0b66edc-73c6-426f-9f7c-3ef1eace5330)

etobella commented:
I applied a fix for this. I am not sure if it gets all the possible cases, but at least it is an approach

florian-dacosta commented:
@etobella Thanks for your work, it is really helpull !
About the rounding case, with last version of the code, here is what I have : 
1) First click on the reconcile button on journal 
![1-start-reconciliation](https://github.com/user-attachments/assets/2b3a02d4-b8de-498e-9255-583505f7ed6c)

2) It give me the reconciliation screen with the rounding error in the suspense account : 
![2-before-begining-reconcile-ui](https://github.com/user-attachments/assets/4475d284-ce1e-4e40-b318-19501c682a08)

If I choose the counter part, all is fine, then when I reconcile, as you saif, all is fine, the invoice is reconciled correctly, the accounting entries are correct.

3) Right after, the view of the statement line in the reconciliation UI was correct. I am not sure how it work, but with a new tab/after restarting the server and going back on the statement line, I have really strange data : 
![3-after-reconciled-view](https://github.com/user-attachments/assets/bed21f81-8074-4f0d-bead-b8a5ade7c29d)

For now the most annoying for me is the display in 1. Even if all goes fine after, it is very confusing for the user.



> I applied a fix for this. I am not sure if it gets all the possible cases, but at least it is an approach

I am not sure what you refer to ?
If it is about the last created issue (695), I have just updated it because I had made a mistake in the screen shot of the use case.
I have tested again right now, and it is still actual.

etobella commented:
1st point should be solved now.

florian-dacosta committed:
[FIX] account_reconcile_oca : Fix multi currency management
Fix the case of payment with a foreign currency set on bank statement line

improve tests around multi-currency

florian-dacosta commented:
> 1st point should be solved now.

It is, thanks
I have made a PR on yours to fix different other issues, Take a look when you can

florian-dacosta commented:
@etobella Could you rebase please ?

yvaucher commented:
/ocabot rebase

yvaucher commented on account_reconcile_oca/models/account_bank_statement_line.py:
```diff
@@ -291,7 +332,9 @@ def _check_line_changed(self, line):
-                self.manual_partner_id and self.manual_partner_id.name_get()[0] or False
+                self.manual_partner_id
+                and self.manual_partner_id.name_get()[0]
+                or [False, False]
```
`bool([False, False])` => `True`

Are you sure you want this?

florian-dacosta replied:
I think we compare this result to [False, False] yes, but I will check again

florian-dacosta replied:
If, from the reconciliation UI, for any bank statement line (which is not linked to any partner) you click on the counter part line, you can see that the onchange is called, and in the values of `reconcile_data_info` field, we have 
`[False, False]` in the `partner_id` key (in `data` list).
Then, if we compare here `False` with `[False, False]`, like it was before, it will be considered as a change where it is actually the same (there is no partner_id !).

That is why I did adapt here but I agree it is not nice and it could be an issue, since Odoo usually put `False` when there are not partner : https://github.com/OCA/account-reconcile/blob/16.0/account_reconcile_oca/models/account_reconcile_abstract.py#L69

But I don't know why the UI send `[False, False]` in the onchange, nor if it could be changed.
Any idea @etobella @yvaucher ?

florian-dacosta replied:
Well, it actually comes from https://github.com/OCA/account-reconcile/blob/17.0/account_reconcile_oca/models/account_bank_statement_line.py#L891

I'll do a commit to propose a change about it.

florian-dacosta replied:
Sorry for the noise here.
I'll make it short here. When there are not partner at all, in the data, the value  `[False, False]` is set here : 
https://github.com/OCA/account-reconcile/blob/16.0/account_reconcile_oca/models/account_bank_statement_line.py#L405
and here : 
https://github.com/OCA/account-reconcile/blob/16.0/account_reconcile_oca/models/account_bank_statement_line.py#L896

So when checking if the line has changed, and there is no partner at all, comparing the reconcile_data partner part (`[False, False]`)with `False` will always be considered different while no changed occured (no partner was set).

So I think we could leave this as is. (meaning comparing` [False, False]` with `line.get("partner_id")`

PabloMartFL reviewed (approved):
Tested the PR on runboat and it resolves the problems in the issues:

- For the test, I created a company in EUR and created a Bank in EUR.
- The currency rate are the following:
- 
![Captura desde 2024-10-04 12-23-36](https://github.com/user-attachments/assets/27dd4947-3117-44d8-bc74-3e50db7967ec)

The suspended balance apears correct and when you select the invoices that fully match with the statement, it doesnt correct the invoice amount.
Also, in the account move, the amount in currency is displayed correcly

https://www.awesomescreenshot.com/video/32216430?key=6538e7535b46717a2905cfcb08d9a052

florian-dacosta committed:
[FIX] account_reconcile_oca : foreign currency reconcile with late currency rate
It is possible that the statement line in foreign currency is created before the rate of the day is updated in Odoo. In this case we need to take the real rate of the statement line to comput the exchange rate

florian-dacosta committed:
[FIX] account_reconcile_oca : max_amount rounding error leading to unwanted currency conversion

florian-dacosta committed:
[FIX] account_reconcile_oca : exchange rate gain/loss currency and amount
If currency amount is not 0, the suspense line will have wrong amount

ljsalvatierra-factorlibre commented:
Hi @etobella can you do a rebase?

It has conflicts with https://github.com/OCA/account-reconcile/pull/758

ljsalvatierra-factorlibre commented on account_reconcile_oca/models/account_bank_statement_line.py:
```diff
@@ -561,16 +612,87 @@ def _default_reconcile_data(self, from_unreconcile=False):
+                        == self.company_id.currency_exchange_journal_id
+                    ):
+                        reconcile_auxiliary_id, lines = self._get_reconcile_line(
+                            reconciled_line.move_id.line_ids - reconciled_line,
```
Expected singleton.

```
>>> reconcile_auxiliary_id, lines = self._get_reconcile_line(
>>> new_vals = super()._get_reconcile_line(
>>> original_amount = amount = net_amount = line.debit - line.credit
>>> record.ensure_one()
```

pedrobaeza reviewed (approved):
Tested on real cases and working.

/ocabot merge patch

etobella committed:
[FIX] account_reconcile_analytic_tag: Add kwargs on _get_reconcile_line

etobella committed:
[FIX] account_reconcile_oca: Update currency_amount

etobella commented:
/ocabot merge minor
